File: regions/system/library/dart/ratelimit_client/test/ratelimit_client_test.dart
import 'package:test/test.dart';
import 'package:k1s0_ratelimit_client/ratelimit_client.dart';

void main() {
  late InMemoryRateLimitClient client;

  setUp(() {
    client = InMemoryRateLimitClient();
  });

  group('check', () {
    test('returns allowed when under limit', () async {
      final status = await client.check('test-key', 1);
      expect(status.allowed, isTrue);
      expect(status.remaining, equals(99));
      expect(status.retryAfterSecs, isNull);
    });

    test('returns denied when over limit', () async {
      client.setPolicy('limited', const RateLimitPolicy(
        key: 'limited',
        limit: 2,
        windowSecs: 60,
        algorithm: 'fixed_window',
      ));

      await client.consume('limited', 2);
      final status = await client.check('limited', 1);
      expect(status.allowed, isFalse);
      expect(status.remaining, equals(0));
      expect(status.retryAfterSecs, equals(60));
    });
  });

  group('consume', () {
    test('consumes and returns remaining', () async {
      final result = await client.consume('test-key', 1);
      expect(result.remaining, equals(99));
      expect(client.getUsedCount('test-key'), equals(1));
    });

    test('throws when exceeding limit', () async {
      client.setPolicy('small', const RateLimitPolicy(
        key: 'small',
        limit: 1,
        windowSecs: 60,
        algorithm: 'token_bucket',
      ));

      await client.consume('small', 1);
      expect(
        () => client.consume('small', 1),
        throwsA(isA<RateLimitError>()),
      );
    });
  });

  group('getLimit', () {
    test('returns default policy', () async {
      final policy = await client.getLimit('unknown');
      expect(policy.limit, equals(100));
      expect(policy.windowSecs, equals(3600));
      expect(policy.algorithm, equals('token_bucket'));
    });

    test('returns custom policy', () async {
      client.setPolicy('tenant:T1', const RateLimitPolicy(
        key: 'tenant:T1',
        limit: 50,
        windowSecs: 1800,
        algorithm: 'sliding_window',
      ));

      final policy = await client.getLimit('tenant:T1');
      expect(policy.key, equals('tenant:T1'));
      expect(policy.limit, equals(50));
      expect(policy.algorithm, equals('sliding_window'));
    });
  });

  group('RateLimitError', () {
    test('contains code and message', () {
      const error = RateLimitError('exceeded', code: 'LIMIT_EXCEEDED', retryAfterSecs: 30);
      expect(error.code, equals('LIMIT_EXCEEDED'));
      expect(error.retryAfterSecs, equals(30));
      expect(error.toString(), contains('LIMIT_EXCEEDED'));
    });
  });

  group('RateLimitStatus', () {
    test('stores all fields', () {
      final status = RateLimitStatus(
        allowed: true,
        remaining: 50,
        resetAt: DateTime.now(),
      );
      expect(status.allowed, isTrue);
      expect(status.remaining, equals(50));
      expect(status.retryAfterSecs, isNull);
    });
  });
}

File: regions/system/library/typescript/ratelimit-client/__tests__/ratelimit-client.test.ts
import { describe, it, expect } from 'vitest';
import { InMemoryRateLimitClient, RateLimitError } from '../src/index.js';
import type { RateLimitPolicy } from '../src/index.js';

describe('InMemoryRateLimitClient', () => {
  it('checkで許可を返す', async () => {
    const client = new InMemoryRateLimitClient();
    const status = await client.check('test-key', 1);
    expect(status.allowed).toBe(true);
    expect(status.remaining).toBe(99);
    expect(status.retryAfterSecs).toBeUndefined();
  });

  it('checkで制限超過を返す', async () => {
    const client = new InMemoryRateLimitClient();
    const policy: RateLimitPolicy = {
      key: 'limited',
      limit: 2,
      windowSecs: 60,
      algorithm: 'fixed_window',
    };
    client.setPolicy('limited', policy);

    await client.consume('limited', 2);
    const status = await client.check('limited', 1);
    expect(status.allowed).toBe(false);
    expect(status.remaining).toBe(0);
    expect(status.retryAfterSecs).toBe(60);
  });

  it('consumeで使用量を消費する', async () => {
    const client = new InMemoryRateLimitClient();
    const result = await client.consume('test-key', 1);
    expect(result.remaining).toBe(99);
    expect(client.getUsedCount('test-key')).toBe(1);
  });

  it('consumeで制限超過時にエラーを投げる', async () => {
    const client = new InMemoryRateLimitClient();
    client.setPolicy('small', {
      key: 'small',
      limit: 1,
      windowSecs: 60,
      algorithm: 'token_bucket',
    });

    await client.consume('small', 1);
    await expect(client.consume('small', 1)).rejects.toThrow(RateLimitError);
  });

  it('getLimitでデフォルトポリシーを返す', async () => {
    const client = new InMemoryRateLimitClient();
    const policy = await client.getLimit('unknown');
    expect(policy.limit).toBe(100);
    expect(policy.windowSecs).toBe(3600);
    expect(policy.algorithm).toBe('token_bucket');
  });

  it('getLimitでカスタムポリシーを返す', async () => {
    const client = new InMemoryRateLimitClient();
    client.setPolicy('tenant:T1', {
      key: 'tenant:T1',
      limit: 50,
      windowSecs: 1800,
      algorithm: 'sliding_window',
    });

    const policy = await client.getLimit('tenant:T1');
    expect(policy.key).toBe('tenant:T1');
    expect(policy.limit).toBe(50);
    expect(policy.algorithm).toBe('sliding_window');
  });

  it('RateLimitErrorにコードが含まれる', () => {
    const err = new RateLimitError('exceeded', 'LIMIT_EXCEEDED', 30);
    expect(err.code).toBe('LIMIT_EXCEEDED');
    expect(err.retryAfterSecs).toBe(30);
    expect(err.name).toBe('RateLimitError');
  });

  it('複数コストのcheckが正しく動作する', async () => {
    const client = new InMemoryRateLimitClient();
    client.setPolicy('cost-key', {
      key: 'cost-key',
      limit: 10,
      windowSecs: 60,
      algorithm: 'fixed_window',
    });

    const status1 = await client.check('cost-key', 5);
    expect(status1.allowed).toBe(true);
    expect(status1.remaining).toBe(5);

    const status2 = await client.check('cost-key', 11);
    expect(status2.allowed).toBe(false);
  });
});
